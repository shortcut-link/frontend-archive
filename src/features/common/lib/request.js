//@flow
import { $token } from '../model/token';
import type { Method, Options, CustomResponse } from './type';

const baseURI = '/api';

export const request = (
  method: Method,
  url: string,
  options: Options = {}
): Promise<any> => {
  const token = $token.getState();

  const headers = new Headers({
    ...createContentType(options),
    ...createAuthorization(token),
    ...options.headers
  });

  const uri = `${options.baseURI || baseURI}${url}`;

  const { body, ...restOptions } = options;

  const config = new Request(uri, {
    method,
    headers,
    ...restOptions,
    body: createBody(options, headers)
  });

  return fetch(config).then(response => {
    if (options.parse === 'text') {
      return response.text;
    }

    if (options.parse === 'noparse') {
      return response.text;
    }

    const contentType = response.headers.get('Content-Type');

    if (contentType && contentType.includes('json')) {
      return response.json().then(responseToPromise, responseToPromise);
    }

    throw new TypeError('Unexpected content-type');
  });
};

const createContentType = (options: Options) => {
  const header = contentTypeFromOptions(options);

  return header ? { 'Content-Type': header } : {};
};

const contentTypeFromOptions = (options: Options) => {
  if (options && options.headers && options.headers['Content-Type']) {
    return options.headers['Content-Type'];
  }

  if (options && options.body && options.body instanceof FormData) {
    return 'multipart/form-data';
  }

  return typeof options.body === 'object' ? 'application/json' : '';
};

const createAuthorization = (token: ?string) =>
  token ? { Authorization: `bearer ${token}` } : {};

const createBody = (
  options: Options,
  headers: Headers
): FormData | string | void => {
  const contentType = headers.get('content-type');

  if (options.body && contentType && contentType.includes('json')) {
    return JSON.stringify(options.body);
  }

  if (options.body instanceof FormData) {
    return options.body;
  }

  return undefined;
};

function responseToPromise<R, E>(response: CustomResponse<R, E>) {
  return response && typeof response.ok === 'boolean'
    ? okToPromise<R, E>(response)
    : response;
}

function okToPromise<R, E>(response: CustomResponse<R, E>): Promise<R> {
  return response.ok
    ? Promise.resolve(response.result)
    : Promise.reject(response.error);
}
