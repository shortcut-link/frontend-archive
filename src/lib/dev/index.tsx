import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';

import { lightTheme } from 'ui/themes';

/**
 * For testing components that require ThemeProvider or BrowserRouter
 */
export const ThemeRouterProvider = (component: React.ReactNode) =>
  renderer.create(
    <ThemeProvider theme={lightTheme}>
      <BrowserRouter>{component}</BrowserRouter>
    </ThemeProvider>
  );
