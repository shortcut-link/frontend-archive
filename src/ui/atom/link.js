import styled from 'styled-components';
import { Link as RouterLink } from 'react-router-dom';

export const Link = styled(RouterLink)`
  text-decoration: none;
  transition: 0.15s;
  color: inherit;

  &:hover {
    color: ${({ theme }) => theme.palette.primary.hover.background};
  }
`;
