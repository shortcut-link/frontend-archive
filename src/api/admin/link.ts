import { request } from 'lib/request';
import { linkError } from '../link.errors';
import { Link, LinkParameter, LinkParameterValue } from '../link';

export interface FoundLink extends Link {
  user: string;
}

const find = (url: string) =>
  request<FoundLink>('GET', `/admin/link/find?url=${url}`).catch(linkError);

const changeParameter = (
  url: string,
  parameter: LinkParameter,
  value: LinkParameterValue
) =>
  request<void>(
    'POST',
    `/admin/link/parameter?url=${url}&parameter=${parameter}`,
    {
      body: { value }
    }
  );

const remove = (url: string) =>
  request<void>('DELETE', `/admin/link?url=${url}`).catch(linkError);

export const adminLinkAPI = { find, changeParameter, remove };
