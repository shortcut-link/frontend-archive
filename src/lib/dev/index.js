import React from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ToggleThemeProvider } from 'lib/theme';
import { lightTheme, darkTheme } from 'ui/themes';

/* For testing components that
 * require ThemeProvider or BrowserRouter
 */

export const ThemeProviderRouter = ({ children }) => (
  <ToggleThemeProvider light={lightTheme} dark={darkTheme}>
    <BrowserRouter>{children}</BrowserRouter>
  </ToggleThemeProvider>
);
