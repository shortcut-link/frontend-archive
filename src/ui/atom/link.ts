import styled, { css } from 'styled-components';
import { Link as RouterLink } from 'react-router-dom';
import { mixins } from 'lib/styled-components';

export const cssLink = css`
  padding-top: 10px;
  padding-bottom: 10px;
  text-decoration: none;
  transition: 0.15s;
  color: inherit;
  font-size: inherit;
  outline: none;

  &:hover,
  &:focus {
    color: var(--primary-hover);
  }
`;

export const Link = styled(RouterLink)`
  ${mixins}
  ${cssLink}
`;
