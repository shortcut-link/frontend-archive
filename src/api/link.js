import { request } from 'lib/request';
import { linkError } from './link.errors';

/**
 * Getting data about the current user session
 * @returns { Promise<{ user: { email: string } }> }
 */
const create = url =>
  request('POST', '/link', { body: { url } }).catch(linkError);

/**
 * Change link options
 * @param {string} id
 * @param { { tracking?: string } } options
 */
const changeUserLinkOptions = (url, options) =>
  request('POST', '/link/options', { body: { url, options } });

export const linkAPI = {
  create,
  changeUserLinkOptions
};
