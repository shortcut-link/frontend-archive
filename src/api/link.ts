import { request } from 'lib/request';
import { linkError } from './link.errors';

export interface Link {
  url: string;
  originalUrl: string;
  transitions?: number;
  createdAt: Date;
}

export interface CreatedLink {
  url: string;
}

const create = (url: string) =>
  request<CreatedLink>('POST', '/link', { body: { url } }).catch(linkError);

const remove = (url: string) =>
  request<void>('DELETE', `/link?url=${url}`).catch(linkError);

export interface LinkOptions {
  tracking?: boolean;
}

const changeUserLinkOptions = (url: string, options: LinkOptions) =>
  request<void>('POST', '/link/options', { body: { url, options } });

const find = (url: string) =>
  request<void>('GET', `/link/find?url=${url}`).catch(linkError);

export const linkAPI = {
  create,
  remove,
  changeUserLinkOptions,
  find
};
