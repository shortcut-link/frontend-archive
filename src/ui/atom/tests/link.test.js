import React from 'react';
import renderer from 'react-test-renderer';

import { Link } from '../link';
import { ThemeProviderRouter } from 'lib/dev';

it('ui link with wrapper router link', () => {
  const component = (
    <ThemeProviderRouter>
      <Link to={'/profile'}>Ссылочка</Link>
    </ThemeProviderRouter>
  );

  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
