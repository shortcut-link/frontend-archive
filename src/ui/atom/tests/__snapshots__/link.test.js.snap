// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui link with wrapper router link 1`] = `
.c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  color: inherit;
}

.c0:hover {
  color: #2196F3;
}

<a
  className="c0"
  href="/profile"
  onClick={[Function]}
>
  Ссылочка
</a>
`;
